lib/conjur/base.rb

```diff
@@ -253,52 +271,139 @@ def with_audit_resources resource_ids
       end
     end
 
+    module TokenExpiration
+      # The four minutes is to work around a bug in Conjur < 4.7 causing a 404 on 
+      # long-running operations (when the token is used right around the 5 minute mark).
+      TOKEN_STALE = 4.minutes
+
+      attr_accessor :token_born
+
+      def needs_token_refresh?
+        token_age > TOKEN_STALE
+      end
+
+      def token_age
+        gettime - token_born
+      end
+    end
+
+    # When the API is constructed with an API, the token can be refreshed using
+    # the username and API key. This authenticator assumes that the token was
+    # minted immediately before the API instance was created.
+    class APIKeyAuthenticator
+      include TokenExpiration
+
+      attr_reader :username, :api_key
+
+      def initialize username, api_key
+        @username = username
+        @api_key = api_key
+        update_token_born
+      end
+
+      def refresh_token
+        Conjur::API.authenticate(username, api_key).tap do
+          update_token_born
+        end
+      end
+
+      def update_token_born
+        self.token_born = gettime
+      end
+
+      protected
+
+      def gettime
+        Process.clock_gettime Process::CLOCK_MONOTONIC
+      rescue
+        # fall back to normal clock if there's no CLOCK_MONOTONIC
+        Time.now.to_f
+      end
+    end
+
+    # When the API is constructed with a token, the token cannot be refreshed.
+    class UnableAuthenticator
+      def refresh_token
+        raise "Unable to re-authenticate using an access token"
+      end
+
+      def needs_token_refresh?
+        false
+      end
+    end
+
+    # Obtains fresh tokens by reading them from a file. Some other process is assumed
+    # to be acquiring tokens and storing them to the file on a regular basis.
+    # 
+    # This authenticator assumes that the token was created immediately before
+    # it was written to the file.
+    class TokenFileAuthenticator
+      include TokenExpiration
+
+      attr_reader :token_file
+
+      def initialize token_file
+        @token_file = token_file
+        update_token_born
+      end
+
+      def refresh_token
+        JSON.parse(File.read(token_file)).tap do
```

Comment:
I know this is just following the old API, but maybe we shouldn't try to parse the token and just treat it as opaque. This would be more future-proof. Even `#authenticate` docstring indicates (erroneously) that it should return a String.